cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/BatchSparkPipelineDriver.java

```diff
@@ -154,8 +154,7 @@ protected SparkCollection<RecordInfo<Object>> getSource(StageSpec stageSpec,
 
       // Return future collection containing the method to read from the SQL engine and a fallback using the Spark
```

Comment:
Since we are not using the `sparkCollectionSupplier` anymore, it might make sense to bring that logic back into this function.

Also, since LazySparkCollection is simply "resolving" the read collection, I think it makes sense to simply return a SQLBackedCollection that is still pending to resolve (in the same way that join tasks take time to process and the BatchSQLEngineAdapter needs to wait for). What do you think?

Reply:
Essentially, directly return the contents of https://github.com/cdapio/cdap/blob/245247f7676c7bb709435a753f21d04abde79040/cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/BatchSparkPipelineDriver.java#L458 (without waiting for the read task to succeed) and if the read task itself fails, the entire pipeline needs to fail.